compiler/next/include/chpl/uast/Range.h

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CHPL_UAST_RANGE_H
+#define CHPL_UAST_RANGE_H
+
+#include "chpl/queries/Location.h"
+#include "chpl/uast/Expression.h"
+
+namespace chpl {
+namespace uast {
+
+
+/**
+  This class represents a range expression. For example:
+
+  \rst
+  .. code-block:: chapel
+
+      // Example 1:
+      var r = 0..15;
+
+  \endrst
+
+ */
+class Range final : public Expression {
+ public:
+  enum IntervalKind {
+    OPEN_BOTH,
+    OPEN_HIGH,
+    OPEN_LOW,
+    CLOSED
```

Comment:
I see (looking at the parser) that unbounded ranges like `1..` or `..10` or `..` are stored as CLOSED which I find intensely confusing. Can this enum also have one or more UNBOUNDED elements? Could have unbounded low / unbounded high / unbounded both or just unbounded.

Actually I think it'd be more pleasant to represent if there is an enum and the Range stores two values; one for the lower part and one for the upper.

```
  enum BoundKind {
    CLOSED, // e.g. 1..10
    OPEN,     // e.g. 1<..<10
    UNBOUNDED, // e.g. ..
  };
  BoundKind lowerBoundKind;
  BoundKind upperBoundKind;
```

Such a representation seems more satisfying, at least to me.

Reply:
All this enum is meant to express is the kind of interval used. Would it help if I stuck the word `INTERVAL` in there? Because it's not really saying anything about the bounds (and neither do the operators in the parser, e.g. `TDOTDOTOPENHIGH`).

```cpp
enum IntervalKind {
  INTERVAL_OPEN_BOTH,
  INTERVAL_OPEN_HIGH,
  INTERVAL_OPEN_LOW,
  INTERVAL_CLOSED
};
```

I don't really like the idea of querying whether a range is unbounded via the enum. All you need to do is detect whether it has no lower/upper bound (the `lowerBound` and `higherBound` methods return `nullptr`). To have both an enum for each bound and accessors that can return nullptr seems like overkill to me (especially when you can detect that a range is unbounded by checking both of either one).

I think @bradcray had similar confusion about this but I can't remember exactly what it was. I'd be curious to see what he thinks.

Reply:
> To have both an enum for each bound and accessors that can return nullptr seems like overkill to me (especially when you can detect that a range is unbounded by checking both of either one).

It doesn't seem like a big deal to me. Based on the enum, the lowerBound / higherBound should or shouldn't be nullptr, and we can ensure this property in the constructor.

Reply:
I am with Michael on this. I don't want to get philosophical and think whether `1..` is an open interval or closed one (if it matters; it is more of an open interval to me because it doesn't include it's bound because there's nothing to include), and would prefer if nonexistence was represented by a different enum value.

As the user of this code, I might choose to check whether bounds are null or query the enum to get the information I need, and would hope that what I get is consistent and something I'd expect. Without representing boundedness in the enum, I don't think that's achievable.

Maybe if we have a new AST type that's `RangeBound`, then it may be considered a binary choice to have it open or closed, and its nonexistence would signify unboundedness. But that's likely overkill.

Reply:
I still don't understand how the interval has anything to do with the bound, other than you can't have an open high interval without a high bound (but that's not even parseable).

Reply:
From purely math sense, a "closed" interval is one that includes both its bounds. I don't think an unbounded range can be closed, because, what is there to include? Still in the math sense, Chapel's unbounded range *could* be interpreted as `1..<inf`, which is still only a "half-closed" interval in mathland.

In any case, I think this confusion is why an unbounded range shouldn't have anything that says it's "closed".

Reply:
Sounds good. I was just treating the interval like a nudge operation. I'll go ahead and make the change.

Reply:
I'm coming to this late, but I'd been thinking of ".." vs. "..<" as one decision that needs to be represented (where David's enum also supported "<.." and "<..<" as future expansions).  And then whether or not the low / high bounds were given or not as a separate decision.  Note that an "unbounded open" range like "..<" can make sense in a slicing context.  For example, we might say `A[..<]` is a way to slice all of A's elements other than its uppermost, for 1D array A (though I don't believe this is supported today... but I think it should be).

So to me, I think you (we) still need a separate way of representing "which range syntax is used to build the range" vs. "which bounds are present/absent orthogonally, whatever you call them.  Put another way, this enum:

```chapel
  enum BoundKind {
    CLOSED, // e.g. 1..10
    OPEN,     // e.g. 1<..<10
    UNBOUNDED, // e.g. ..
  };
```

doesn't seem to represent siblings to me since, if OPEN can be missing bounds, it seems that CLOSED should be able to as well.

Reply:
If I were the only developer, I'd focus less on the bounds and more on the operator used to create the range (`..`, `..<`):

```chapel
enum RangeOp {
  NORMAL,     // `..`
  OPEN_UPPER  // `..<`
}
```

Where I'd reserve OPEN_LOWER and OPEN_BOTH for `<..` and `<..<` if/when we add them.  Or, I'd just use a boolean to select between the normal and open cases we support today and switch to an enum if/when we added the other two open interval types.

Reply:
I am thinking about adjusting the `enum` to be called `OpKind` and it will contain:

```cpp
enum OpKind {
  DEFAULT,
  OPEN_UPPER,
  OPEN_LOWER,
  OPEN_BOTH
};
```

Hopefully this does not have any semantic implications. Does this seem OK to everyone? I can also use `NORMAL` instead of `DEFAULT`.

Reply:
That seems OK to me for now, but only because we don't have unbounded ranges like `..<` today. (Once we have those I will again be confused by how an "unbounded open" range is as different from any other unbounded range; but we can worry about that in the future).

Reply:
Would choosing a different name for the ops besides `OPEN_...` help in that regard? I'm struggling to think of anything, though.

Reply:
The builder calls internal functions e.g. `nudgeUpper` `nudgeLower`, maybe we can use something like that?

Reply:
Yes, other terms besides `open` would help me if we do support those, that's why I proposed e.g. using `HAS_LESS_THAN` or something (but maybe nobody liked that).  `nudge` doesn't mean anything to me so that'd be more about defining our own term (which would also be OK).

Reply:
How about...

```cpp
enum OpKind {
  DEFAULT,
  UPPER_LESS_THAN,
  LOWER_LESS_THAN,
  BOTH_LESS_THAN
};

Reply:
@bradcray @e-kayrakli Do these enum values seem sufficient to you?

Reply:
I personally find using "less than" clunky here in its focus on the typography rather than the meaning, and maybe even slightly inaccurate / confusing in that I think of the `<..` range constructor as implying more of a "greater than" relationship than a "less than" relationship since the indices we're focused on are to the right of the `<` symbol.  But I'm also pretty much beyond caring at this point (until I have to start working in the new compiler front-end at least) given that my suggestions aren't getting any love and most of the problematic cases are ones that aren't supported (i.e., we don't have a `<..` operator yet; we don't have a `lo..<` form yet).

I went to look at what we call things in the spec, and there we use "closed interval ranges" and "half-open interval ranges", so arguably the spec needs modifications as well to counter Michael and Engin's concern that we shouldn't have closed interval ranges that are unbounded on one or both sides.  If we believed the spec was appropriate / accurate, I would follow its terminology in the code as well.

In lone developer mode, if I had to sprint to an answer based on this conversation, I'd do `bool halfOpenInterval;` where `false` => `..` and `true => `..<` with an assertion that the upper bound can't be elided if it's `true`.

Reply:
>  But I'm also pretty much beyond caring at this point (until I have to start working in the new compiler front-end at least) given that my suggestions aren't getting any love

I don't think that's true, if your suggestions weren't getting any love then we'd still be doing the one-enum-value-per-bound approach :).

> most of the problematic cases are ones that aren't supported

These things already have rules in the parser (albeit they are commented out). This suggests/reinforces that we can just have an enum with two values for now. That seems easier to extend in the future as opposed to a `bool` (which is why I'm opting away from it), if we ever decide to support OPEN_LOW or OPEN_BOTH.

> I personally find using "less than" clunky here in its focus on the typography rather than the meaning

I agree, I'm also personally not satisfied with it.

Given that feedback, I think I'm just going to go with your suggestion of:

```cpp
enum OpKind {
  DEFAULT,
  OPEN_UPPER
};
```

For now. Does that seem reasonable?

@mppf Would you be OK with this enum if it only has two values (for the foreseeable future)? We're back to punting on what `OPEN_LOWER` and `OPEN_BOTH` mean in the future.

Reply:
Yep that is OK with me (and I arguably already said that in https://github.com/chapel-lang/chapel/pull/17975#discussion_r664826727 ). We can change it in the future if needed (I don't think we're aiming for stability in these APIs right now - just trying to have a design the team is happy with).